### ISSUE TITLE
"g" goto does not work anymore

### ISSUE THREAD

**Issue by internationils** [opto/nostalgy-xpi#137]
When using "g" to go to a folder, it brings up the field and selection, but when selected / enter is hit, nothing happens (no folder change).
- 91.2.1 (64-bit) on current Ubuntu / 3.0.4
- 91.3.0 (64-bit) on Win7 / 3.0.4

**Comment by cjholly** [comment]
In Windows at least, I changed the key assignment from G to Control-G and it works fine. I think TB assimilated G for its own . . .  --Chris

**Comment by internationils** [comment]
The weird thing is that the "g" key works... I can select the folder, but when hitting Enter it does not go there, it remains in the current folder.

**Comment by Bathsheba** [comment]
I still see this, it's a dang nuisance!  Refining it: g works if the folder name is visible in the Folder Pane, but not if it's hidden.

I groveled through the whole list of T-bird hotkeys and did not find g used as a shortcut natively, so if it's an issue of competition for the key then it's not an obvious one.

**Comment by opto** [comment]
There are some spurious things happening where initialisation comes at the wrong point in time, because TB 91 does not give addons a clear signal (yet) when it is finished initialising. So Nostalgy may initialise and later TB overwrites that.

I just tried G on Win 11, TB 91.4.0 and - unfortunately - it works as expected. That is what makes this (and some other) errors so difficult to debug.

My current idea is to put a 1 second delay before our initialisation. And see whether that helps. I will upload that ca. tonight.

Klaus

**Comment by Bathsheba** [comment]
tyvm.  I'm Windows 10, TB 91.4.0

**Comment by Olifair** [comment]
Well, just to also leave a positive return : I am also running Windows 10, I have had all versions of TB, up to now 91.4.0 and using Nostalgy++ 3.0.8 and I have never had any problems with the G, neither just "G" nor SHIFT-G, so it seems to me that these problem must be installation specific, either due to other installed TB extensions, or due to the problem that opto is mentionning. 
Cheers
Oliver

**Comment by internationils** [comment]
I can confirm what Bathsheba said: if the folder is visible (can be on-screen or off-screen) in the folder pane, then "g" works, if it is hidden (no new mails in the unread view, or collapsed) then it (sometimes) works. It's actually not that simple and weirder than that though, even if I un-collapse my entire mail folder tree, I still cant jump to some folders. If I scroll to them to see them on screen and then "g", still no-go. If I click them and then collapse them, go to another folder in the same server, I can get back to it. If I go to another mailserver while leaving them uncollapsed, I can't get there. I can't 100% figure out when it does and does not work, but it seems like there is some list that does not contain all folders so it can't jump there, or it is unwilling to jump to folders that are not "active" in some way (uncollapsed or recently selected. Either way, its weird and used to work fine a few versions ago.
I have the same behavior on an Ubuntu box with TB 91.3.1 and a Win 11 laptop with TB 91.4.0 ... extensions are minimal, copy-sent-to-current, automatic-dictionary, nostalgy, and folder pane view switcher (disabled because incompatible) on both systems.

**Comment by Olifair** [comment]
I have no problem to use g or Shift-G to go to any folder I like, collapsed or not. And this works for me as well for local folders as it does for at least second order folder on different IMAP servers I use.  If your problem is with IMAP servers, perhaps it is a problem with the way the folder structure is synched. In the TB IMAP server advanced preferences I have checked "taking into account folders which contain sub folders" (wording might be different, I'm translating), and "Allow server to further than these space names" (I have difficulties to translate, it's the last option at the bottom). I do not have checked "show only folders with a subscription".  Hope this helps. Cheers.

**Comment by internationils** [comment]
"Server supports folders that contain sub-folders and messages! ... that is checked. Checking or unchecking the "show only folders with a subscription" does not make a difference. I will try uninstalling and reinstalling the addon on Monday maybe.

**Comment by Bathsheba** [comment]
@Olifair I set all options as you describe, uninstalled and reinstalled Nostalgy, re-started Tbird, and I still cannot 'g' to a hidden folder.  
I must tell you: this bug is real.

**Comment by Olifair** [comment]
@internationils @Bathsheba  What Nostalgs++ version are you actually using ? 
I have tried what internationils is describing and I do not have any problem at all. Even old folders, not used for ages, can be jumped to. 
Do you see the problem only for G, or also for S and B ?

**Comment by internationils** [comment]
"s" definitely works. "b" only moves, "G" also does not work.
Something else that I noticed is that sometimes I have to click on a folder twice to get it to open... I click on it, and it does not open (even after repeated clicking), I click on another folder and then that folder and I can get to it.

**Comment by Bathsheba** [comment]
> @internationils @Bathsheba What Nostalgs++ version are you actually using ? I have tried what internationils is describing and I do not have any problem at all. Even old folders, not used for ages, can be jumped to. Do you see the problem only for G, or also for S and B ?

I have v3.0.11 . I don't see the problem for S, the message is saved to the hidden folder.  B saves the message to the hidden folder but doesn't take me there.

**Comment by internationils** [comment]
something else weird. After reinstalling 3.0.11 yesterday, TB slowed to a crawl when clicking on an email. Processor usage went up, and it took 30-40 seconds for the mail to come up. Deleting was instantaneous, loading the next one again took that long. Deinstalling nostalgy++ fixed it, reinstalling made the problem show up again. This was only on my Ubuntu box (5.8 kernel), not on the windows box. Any ideas??

**Comment by dstango** [comment]
I can confirm that being on version 3.0.4 selection of a folder that's not displayed after pressing "G" doesn't jump to the selected folder. 
I usually have selected to view "unread folders" and "compact view". Only workaround so far is to view "all folders" - then the jump does work.

**Comment by opto** [comment]
slow operation might be caused by the discovery of  threads/conversations for move destination suggestion. (3.0.11). I am considering to add an option to switch that off, so it can be tested.

Klaus

**Comment by opto** [comment]
you could also try G if set to offline. I experience in my TB that there seems to be now much more requests to the server as before.
(often, my windows mailserver goes to 50%CPU for long time, and I see in the protocol that TB is iterating all folders to retrieve tags/new emails etc.).

**Comment by internationils** [comment]
> slow operation might be caused by the discovery of threads/conversations for move destination suggestion. (3.0.11). I am considering to add an option to switch that off, so it can be tested.
@opto This would be great, I'd be happy to test. Else could you add a "heavy debug" mode that logs activities in detail (with good resolution timestamps) to see what is causing the high load? Thanks & a happy new year!

**Comment by internationils** [comment]
Just tried installing 3.0.11 on TB 91.3.1+build1-0ubuntu0.21.10.2 / Kernel 5.13.0-23-generic #23-Ubuntu SMP Fri Nov 26 11:41:15 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux again, and the slowdown still happens. It happens when I click on a mail... I get the "Loading Message..." info at the bottom in the status bar, and then it literally takes a minute (~60 sec) to load and display the message.
When I am offline and I switch messages, the delay also happens and takes the same amount of time.
It would be great to be able to see some kind of a debugging log to figure out what is happening.

**Comment by internationils** [comment]
@opto I just tried 3.0.6, .8, .10 and .11 ... I can't jump to hidden folders in any of these versions. This includes collapsed folders, but if I open the collapsed folders and then re-collapse I can get to all (most?) of them in that sub-tree.

The hanging on selecting mails starts in .10 (see my comment in https://github.com/opto/nostalgy-xpi/issues/134 )

**Comment by opto** [comment]
I can confirm that in 91 with the new multiple views, G does not work when the folder is not in the view.

e.g.: set view to favorites only, then G to a folder that is not displayed. -> fails.

I just had a look at the code: Nostalgy tries to find a view where the folder is displayed. But due to the changges with multiple views, it fails. So this is confirmed as a bug. I hve time later in january and will fix this then.

current workaround: select all folders for view.

Klaus

**Comment by internationils** [comment]
But it's more than that... the "all folders view" must be selected, and also the folders must have been un-collapsed. If I just select the view I can't select some folders, when I un- and then re-collapse them then I can. TB seems to have some concept of "visible & seen" folders under the hood.

**Comment by opto** [comment]
there maybe some 'hope' for the G problem: major parts of that code are error prone due to the change of TB to multiple views (like all, recent, favorites) etc.

Alain originally did a major rewrite of selectfolder because that had a major bug (long time ago). Not sure whether that bug still  exists, but his rewrite can't work under the new multi mode view.

I am rewriting folderpane view switcher, so i due time, this should also work in Nostalgy++.

**Comment by internationils** [comment]
Excellent! I miss folderpane view switcher... the latest version from September 2021 only works up to TB 90. If you want help testing either, please let me know. I can test on Win11 and Ubuntu.

**Comment by opto** [comment]
well, Alain's workaround is still needed, even if in a modified form: TB's native setFolder is supposed to make the folder visible, but it does not, If its parent is collapsed and invisible, it does not function (

although TB states in ts code: 
'This function will also ensure
   * that the folder is actually being displayed (that is, that none of its
   * ancestors are collapsed.'

That remark in TB's code seems to be outdated. Therefore, no really fast solution is available, but programming by hand.

**Comment by opto** [comment]
to be more specific, the TB build in version seems to work if it forces the view to switch to all (if we tell it to do a 'forceselect'.) Is that ok for you? So if you were at favorites, you would be forced into all view

**Comment by internationils** [comment]
No, unfortunately that does not work. I have a lot of archived folders, and a few (20-30) that mails shows up in, which I see in the "unread" view. What used to happen is that G ended up showing the part of the tree that I jumped to in the unread view, which was great.

As a workaround in the interim, can you force-switch to the "all" view, switch to the folder, and then switch back to the unread view automatically? Or would that force another jump automatically?

Another alternative... if you do need to use the unread view for this, is there a way to collapse / uncollapse folders in such a way that the unread plus the switched to folder are uncollapsed? This is kind of simulating the unread view...

**Comment by dstango** [comment]
Here's another idea for a workaround:
Take a message from the goto folder, mark that message unread, change to the folder, mark the previously toggled message read.

**Comment by opto** [comment]
what would be easy: have go to open the folder in  in all folders view, have ctrl-G (or similiar) revert to your previous view (like unread)

**Comment by opto** [comment]
switching back and forth between views: this gets even more complicated due to the fact that we can now have multiple views open. And in there, a vie can be on top, or furher down (and maybe invisible if the view on top has many folders).

At the moment, for folder pane view switcher, I just implemented a single view mode for switching, just as we had before 91. Otherwise, we also have to take care which view is on top, which at the bottom of the sidebar

**Comment by Bathsheba** [comment]
I feel like if it is impossible to make this work as it did before, the only way g can continue to be useful is if there's a keyboard toggle between All folders and Favorite folders.   Thanks for your consideration!  I see this is a real hairball.

**Comment by opto** [comment]
I would toggle it between All folders and the place you were before going there. So it could toggle back to recent, favorites, unread etc., wherever you came from

**Comment by internationils** [comment]
If that's what it takes, that would work for me. Maybe also add a ctrl-something + A-U-N-F-R option (these are the 5 views) to switch to that view... that way you can G to the all view, and then ctrl-something+U to go to unread afterwards (or ctrl-something+B for Back to previous view)... could that be done?

**Comment by opto** [comment]
this https://github.com/opto/nostalgy-xpi/releases/tag/v3.1.4  is a workaround:

G shows the folder in all view
alt-G goes back where you came from, including setting back the view.

Does that help at this moment?

**Comment by internationils** [comment]
It changes the view, but does not go to the folder... it just stays in the folder where it currently is for me. Both on Ubuntu and Win11.

**Comment by opto** [comment]
Itried as a fresh install (Win 11). What happens if you restart TB?

**Comment by internationils** [comment]
I restarted both TBs... no change unfortunately.

**Comment by opto** [comment]
interesting...
so what I did for testing:
goto all, decide which folder to show
goto recent, deselect all in the menu with the three dots
select a folder in recent
G .. to the one chosen in all
->it goes there, and alt-G goes back.

so what do you do differently, or is it just again a race condition or something strange?
setting the folder and the view upon G is done by an inbuild fuction of TB which claims to always (!) make the folder visble.

maybe the parent of the folder you want to show is hidden?

**Comment by internationils** [comment]
I have All and Unread selected, I am in one of the Unread folders, I "g" to a folder that is collapsed in the All view, and the view changes to All but the folder does not switch (maybe you need to disable the "Unread" view?). Doing "g" again restores back to the All and Unread views selected.

**Comment by opto** [comment]
this solves a race condition between old legacy code and new webextension code and detecting whether we are in a mail tab.
This may or may not solve your problem, but it is worth a try.
https://github.com/opto/nostalgy-xpi/releases/tag/v3.1.6

Klaus

**Comment by internationils** [comment]
Quick check, the tag is 3.1.6, the filename is 3.1.5 ... is that correct? Trying it now.

**Comment by internationils** [comment]
With 3.1.5 no change from 3.1.4... it toggles between the unread and unread+all views with every "g" command on both Ubuntu and Win11.

**Comment by opto** [comment]
sorry wrong tag, 3.1.5 is in the manifest. Sorry to hear it still hangs.

Already Alain Frisch writes in the code that the inbuild TB functions are unreliable, that still seems to hold.

He build a workaround that is obsolete now because it uses TB functions that have been removed. So I have to see how that can be remodeled.

**Comment by internationils** [comment]
Is there maybe a way to do folder "jumps" ? It seems that the issue happens when I go deeper into the folder tree. Maybe jumping only one or two levels at a time... i.e. if going to /work/2021/projects/design from /somewhere_else go to work then projects then design (or maybe even all 4?) And along with that, let the "g" select the ALL view (forget about jumping views for now). Maybe we can iterate towards a solution.

**Comment by internationils** [comment]
Can you jump to a top level folder ( /work in the above example), then force-uncollapse / show the tree at that point, then jump one level lower, and repeat until the destination?

**Comment by opto** [comment]
>Is there maybe a way to do folder "jumps"

well yes, but ...

this is what TB promises to do:
>
```
/**
   * Selects a given nsIMsgFolder in the tree. 
     **This function will also ensure
    that the folder is actually being displayed (that is, that none of its
    ancestors are collapsed.**
   *
   * @param aFolder  the nsIMsgFolder to select
   * **@param [aForceSelect] Whether we should switch to the default mode to
   *      select the folder in case we didn't find the folder in the current
   *      view. Defaults to false.**
   * @returns true if the folder selection was successful, false if it failed
   *     (probably because the folder isn't in the view at all)
   */
  selectFolder(aFolder, aForceSelect = false) {
<
```

the two things apparently do not work.:
1) This function will also ensure
    that the folder is actually being displayed (that is, that none of its
    ancestors are collapsed
2 )param [aForceSelect] Whether we should switch to the default mode to
         select the folder in case we didn't find the folder in the current
         view.

**Comment by internationils** [comment]
So I'm assuming that you've tried aForceSelect=true and that the function returns True to indicate that it should have been successful? If so, could you file a TB bug at https://bugzilla.mozilla.org/home ? I've searched there for selectfolder, afolder and aforceselect and not found any filed bugs. A search for https://bugzilla.mozilla.org/buglist.cgi?quicksearch=nsIMsgFolder doesn't show anything that seems relevant to this issue either.

**Comment by internationils** [comment]
What seems to work:
- go to the All Folders and select a folder
- "g" to a folder (will not work but selection of views may change)
- "g" to that folder again, then it goes there

Maybe another workaround idea... select only the "all" view, jump to the Inbox there, then execute the "g" ?

**Comment by internationils** [comment]
To confirm, the above seems to work. Going to the All view and then using "g" twice... the second time it goes there. When trying from the Unread view, I don't think it ever switches over to the All tree.

**Comment by internationils** [comment]
I've filed https://bugzilla.mozilla.org/show_bug.cgi?id=1755446 ... @opto, if there is anything that you can add to the Mozilla bug to be more precise about what nostalgy is trying to do and how, that would be great. Thanks!

**Comment by opto** [comment]
good idea. Because I am already forcing the all view - even if it doesn't work

**Comment by opto** [comment]
Both folder tree and message tree use a `<tree>` component,
which just now is being replaced by something else (I think already in 98).

So any changes in Nostalgy++ code (including the go problem) will/may be obsolete in September, as the underlying TB code will disappear and will be replaced with something else (which might be good, better, or not)

**Comment by Bathsheba** [comment]
Is there any hope here?  I still cannot 'g' to a hidden folder.
It toggles between Favorite and All folders, but folder I want is usually nested 3 levels down.  And of course because it's a toggle, 50% chance that it has turned _off_ All folders, taking me farther from my destination.

Thanks, Bathsheba

**Comment by opto** [comment]
the change in foldertree has been postponed to 114. So yes, I will need to look into this when possible timewise.

Thanks, Klaus

**Comment by Bathsheba** [comment]
Thank you, I shall wait in hope

**Comment by internationils** [comment]
This now works for me in TB102.4.0 with Nostalgy++3.6.1
It switches to the ALL FOLDERS view, and goes to the desired folder. With the "folder pane switcher" addon to switch back to the Unread folders this solution is good enough for me. Thanks or implementing, feel free to close!

**Comment by Bathsheba** [comment]
I'm happy for you!  I don't see any change.

**Comment by internationils** [comment]
Strange. I had to reinstall TB on a new laptop because my Mobo died on the old one, maybe try a clean installation on a different machine? I also see that 3.6.5 is out in https://github.com/opto/nostalgy-xpi/releases ... maybe try that? TB addons is still at 3.6.1.

**Comment by dstango** [comment]
jumping to a hidden folder doesn't work for me, if it's a sub-sub-folder that is not expanded in the all folders view (just tried with 102.4.0). 
So I wouldn't suggest to close the issue.

**Comment by Bathsheba** [comment]
Yes this is my condition also.  As noted above, we are waiting for underlying change and it's not in the control of our honorable dev.

**Comment by internationils** [comment]
Agreed, do not close. I had some situations where it didn't work with deep folders too now.
But it does work better than it used to, before it would almost never jump to anything.

**Comment by internationils** [comment]
Now at TB115.3 and N4.0.5 and the issue is still there. Is it possible with the new underlying code to have a look at this again? The issues seems to be worse since 115/4.0

**Comment by Bathsheba** [comment]
On my computer where I don't use Favorite Folders, "g" can go to any folder.
On my other computer where I do use Favorites, "g" can go to any folder that is visible in the folder pane. It can go to Favorite folders whether or not they are visible. But if I "g" to a folder that is neither Favorite nor visible, the displayed folder does not change, and instead it toggles View/Folders/All. 
Both computers are current version Windows 11 and Thunderbird.
Thanks for looking!

**Comment by opto** [comment]
yes, the folder view switching code used up to now has been removed by TB team, I need to see in TB code how they do it now and then re-implement it
